libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp

```diff
@@ -0,0 +1,453 @@
+
+#include "GetMeterUsingTatumQuantizationFit.h"
+#include "IteratorX.h"
+#include "MirAudioReader.h"
+#include "MirDsp.h"
+#include "MirUtils.h"
+#include "MusicInformationRetrieval.h"
+
+#include <array>
+#include <cassert>
+#include <cmath>
+#include <map>
+#include <numeric>
+#include <regex>
+#include <unordered_map>
+#include <unordered_set>
+
+namespace MIR
+{
+namespace
+{
+constexpr auto minTatumsPerMinute = 100;
+constexpr auto maxTatumsPerMinute = 700;
+constexpr auto minBpm = 50.;
+constexpr auto maxBpm = 200.;
+constexpr auto minBeatsPerBar = 2;
+constexpr auto maxBeatsPerBar = 4;
+constexpr std::array<int, 3> possibleBeatsPerBar { 2, 3, 4 };
+constexpr std::array<double, 9> possibleTatumsPerBeat {
+   1.,                             //
+   2. / 1, 3. / 1, 4. / 1, 6. / 1, //
+   1. / 2, 1. / 3, 1. / 4, 1. / 6  //
+};
+
+// Number of bars and beats per bar dividing the input audio recording.
+struct BarDivision
+{
+   const int numBars;
+   const int beatsPerBar;
+};
+
+constexpr bool IsRound(double x)
+{
+   return static_cast<int>(x) == x;
+}
+
+static_assert(IsRound(1.));
+static_assert(!IsRound(2.5));
+
+// A map of possible number of tatums to a list of number of bars and beats per
+// bar which explain it.
+using PossibleDivHierarchies =
+   std::unordered_map<int /*num tatums*/, std::vector<BarDivision>>;
+
+// Gather a collection of possible number of divisions based on the assumption
+// that the audio is a loop (hence there must be a round number of bars) and on
+// reasonable bar and tatum durations.
+PossibleDivHierarchies GetPossibleDivHierarchies(double audioFileDuration)
+{
+   constexpr auto minBarDuration = 1.;
+   constexpr auto maxBarDuration = 4.;
+   const int minNumBars =
+      std::max(std::round(audioFileDuration / maxBarDuration), 1.);
+   const int maxNumBars = std::round(audioFileDuration / minBarDuration);
+   PossibleDivHierarchies possibleDivHierarchies;
+   for_each_in_range(
+      IotaRange { minNumBars, maxNumBars + 1 }, [&](int numBars) {
+         const auto barDuration = audioFileDuration / numBars;
+         const auto minBpb = std::clamp<int>(
+            std::floor(minBpm * barDuration / 60), minBeatsPerBar,
+            maxBeatsPerBar);
+         const auto maxBpb = std::clamp<int>(
+            std::ceil(maxBpm * barDuration / 60), minBeatsPerBar,
+            maxBeatsPerBar);
+         for_each_in_range(
+            IotaRange { minBpb, maxBpb + 1 }, [&](int beatsPerBar) {
+               std::for_each(
+                  possibleTatumsPerBeat.begin(), possibleTatumsPerBeat.end(),
+                  [&](double tatumsPerBeat) {
+                     const auto tatumsPerBar = beatsPerBar * tatumsPerBeat;
+                     const int numTatums = tatumsPerBar * numBars;
+                     const auto tatumRate = numTatums / audioFileDuration * 60;
+                     if (
+                        IsRound(tatumsPerBar) &&
```

Comment:
I wonder about this use of `IsRound` sometimes wrongly rejecting a hypothesis.  Would it be better to do it with integer operations?

Suppose `tatumsPerBeat` contains nums and denoms,

`(numTatums * num) % denom == 0`

is the condition